feat: Create utility for validating TLS certificates

kfa-aguda opened the pull request:
The first in a series of PRs - this one simply validates that there exists a known TLS peer whose signing certificate's subject matches the agreement certificate's issuer. The PeerInfo returned by this validation is passed upstream to inbound managers/handlers (instead of blindly reading the peer's info directly from the socket).

Closes #11762

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
```diff
@@ -380,8 +385,47 @@ public static boolean hasAnyCauseSuppliedType(
+            final Certificate[] certs = sslsocket.getSession().getPeerCertificates();
+            final X509Certificate agreementCert = (X509Certificate) certs[0];
+            final Optional<PeerInfo> peerOptional = peerInfoList.stream()
+                    .filter(peerInfo -> ((X509Certificate) peerInfo.signingCertificate())
+                            .getSubjectX500Principal()
+                            .equals(agreementCert.getIssuerX500Principal()))
+                    .findFirst();
```
I think we should handle the case where multiple peers match

kfa-aguda replied:
I understand it's sufficient to just return the first match. What would we do with the rest if more than one?

lpetrovic05 replied:
discussed: log when multiple matches

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
```diff
@@ -380,8 +385,47 @@ public static boolean hasAnyCauseSuppliedType(
+                            .equals(agreementCert.getIssuerX500Principal()))
+                    .findFirst();
+            if (peerOptional.isEmpty()) {
+                logger.warn(
+                        "Handshake with client {}:{} was successful but we couldn't find a matching peer. Closing connection.",
+                        sslsocket.getInetAddress(),
+                        sslsocket.getPort());
+            } else {
+                matchingPair = peerOptional.get();
+            }
+        } catch (final SSLPeerUnverifiedException e) {
+            logger.warn(
+                    "Attempt to obtain certificate from an unverified peer {}:{} threw exception {}",
+                    sslsocket.getInetAddress(),
+                    sslsocket.getPort(),
+                    e.getMessage());
+        }
```
logs unfortutanely wont make it to the file without markers

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -90,7 +94,10 @@ private void listen(final ServerSocket serverSocket) throws InterruptedException
             try {
                 final Socket clientSocket = serverSocket.accept(); // listen, waiting until someone connects
-                incomingConnPool.submit(() -> newConnectionHandler.accept(clientSocket));
+                incomingConnPool.submit(() -> {
```
I think we want to avoid having code inside lambdas where possible

kfa-aguda replied:
wrapped in its own class now.

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -102,4 +109,26 @@ private void listen(final ServerSocket serverSocket) throws InterruptedException
+    /**
+     * Adds a handshake completed listener
+     * */
+    private void addHandshakeListener(final Socket clientSocket) {
```
this should be used for identification, but right now it just closes the socket if it cant match a peer

kfa-aguda replied:
This is the expected behavior by my understanding. Do we want to leave the client socket open when we can't match it to a peer?

lpetrovic05 replied:
we want to close the socket, that part is fine. the issue is that it should be used for identifying the peer instead of the current method or reading the public key from the socket

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
```diff
@@ -380,8 +385,47 @@ public static boolean hasAnyCauseSuppliedType(
+                            .equals(agreementCert.getIssuerX500Principal()))
+                    .findFirst();
+            if (peerOptional.isEmpty()) {
+                logger.warn(
```
Should these logs be `warn`? Seems severe enough for `error` IMO

kfa-aguda replied:
someone else connecting to a server socket and getting closed out due to not being a 'valid' connection should not be an error (it doesn't lead to a fault in the app). I think it's ok to warn and move on.

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -48,32 +52,32 @@ public class ConnectionServer implements InterruptableRunnable {
             final SocketFactory socketFactory,
-            final Consumer<Socket> newConnectionHandler) {
+            final Consumer<Socket> newConnectionHandler,
+            final List<PeerInfo> peerInfoList) {
```
can this be annotated `@NonNull`?

kfa-aguda replied:
It can but I opted for consistency - either we annotate all as @NonNull or leave as-is. I opted for as-is.

alittley replied:
IMO there is no problem with inconsistency in this regard.

We want to move toward every parameter having an annotation. If annotating all existing parameters is outside the scope of a PR, at least annotating newly introduced parameters moves us closer to our final goal, rather than building more tech debt to be paid later

kfa-aguda replied:
That's acceptable. Added the annotation to just that one new parameter.

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -102,4 +109,26 @@ private void listen(final ServerSocket serverSocket) throws InterruptedException
+    /**
+     * Adds a handshake completed listener
+     * */
+    private void addHandshakeListener(final Socket clientSocket) {
```
missing nullity annotation for `clientSocket`

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -102,4 +109,26 @@ private void listen(final ServerSocket serverSocket) throws InterruptedException
+     * Adds a handshake completed listener
+     * */
+    private void addHandshakeListener(final Socket clientSocket) {
+        if (clientSocket instanceof final SSLSocket sslsocket) {
```
`sslsocket` -> `sslSocket`

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -102,4 +109,26 @@ private void listen(final ServerSocket serverSocket) throws InterruptedException
+                    try {
+                        event.getSocket().close();
+                    } catch (final IOException e) {
+                        logger.warn(
```
Missing log marker here too

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -102,4 +109,26 @@ private void listen(final ServerSocket serverSocket) throws InterruptedException
+                    try {
+                        event.getSocket().close();
+                    } catch (final IOException e) {
+                        logger.warn(
```
Should this be `error` instead of `warn`?

kfa-aguda replied:
same as the other, i think warn is ok.

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -107,19 +99,6 @@ public void handle(final Socket clientSocket) {
             dis = new SerializableDataInputStream(sis);
             dos = new SerializableDataOutputStream(sos);
 
```
The version check was removed here, but I don't see a replacement anywhere. Are we wanting to get rid of the version check entirely?

kfa-aguda replied:
yes, we're getting rid of it in both the inbound and outbound handlers as the version check is done elsewhere and the previous flag to do them here is always disabled.

alittley commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/connectivity/SocketFactoryTest.java:
```diff
@@ -154,17 +172,47 @@ void tlsFactoryTest(final AddressBook addressBook, final Map<NodeId, KeysAndCert
+                .thenReturn(mock(X500Principal.class));
+
+        final List<PeerInfo> peerInfoListNonMatchingCertPeers = List.of(peer3);
+        testSocketsBoth(
```
I'm finding this test logic difficult to follow, with `testSocketsBoth` asserting outcome of `validateTLSPeer` based on what is or isn't mocked. Maybe there is a way to simplify these assertions / make it clear what the expected outcome is

alittley replied:
Maybe having 2 versions of `testSockets` might make sense: 1 that asserts successful validation, 1 unsuccessful. Or `testSockets` could be broken into separate pieces so that there isn't any duplication.

At a minimum, something to aid understanding would be adding a comment or 2 explaining the logic of checking if an item is mocked

kfa-aguda replied:
Added a comment.

alittley reviewed (commented):
FYI, correct branch naming format is `<issue_number>-branch-description`

kfa-aguda commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
```diff
@@ -380,8 +385,47 @@ public static boolean hasAnyCauseSuppliedType(
+     *      matches the agreement certificate's issuer
+     *
+     * @param sslsocket
+     * 		the address book to create the list from
```
Suggestions?

kfa-aguda replied:
Is 
"Validates that the client on the other end of the socket possesses an agreement
     *  certificate, that the certificate's issuer matches the signing certificate's subject
     *  for one or more known peers"  
better?

alittley replied:
Apologies, I should have been more precise

The issue is with the doc for `sslsocket` which reads "the address book to create the list from", which isn't applicable.

kfa-aguda replied:
thanks, fixed!

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/communication/handshake/HandshakeCompletedListenerImpl.java:
```diff
@@ -0,0 +1,60 @@
+    private final List<PeerInfo> peerInfoList;
+    private static final Logger logger = LogManager.getLogger(HandshakeCompletedListenerImpl.class);
+
+    public HandshakeCompletedListenerImpl(@NonNull final List<PeerInfo> peerInfoList) {
```
Missing javadoc

alittley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/communication/handshake/HandshakeCompletedListenerImpl.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    @Override
+    public void handshakeCompleted(final HandshakeCompletedEvent event) {
```
missing nullity annotation for `event`

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
```diff
@@ -380,8 +388,50 @@ public static boolean hasAnyCauseSuppliedType(
+     * 		List of peers to validate
+     * @return info of the corresponding peer
+     * */
+    public static PeerInfo validateTLSPeer(
```
rename to identifyTlsPeer

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -139,31 +109,48 @@ public void handle(final Socket clientSocket) {
+
+    private PeerInfo getConnectedPeer(@NonNull final SSLSocket sslSocket, @NonNull final List<PeerInfo> peers) {
+        final PeerInfo peer = Utilities.validateTLSPeer(sslSocket, peers);
+        if (peer == null) {
```
add logging for null peer, include cert info

kfa-aguda replied:
done

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -139,31 +109,48 @@ public void handle(final Socket clientSocket) {
+                sslSocket.close();
+            } catch (final IOException e) {
+                logger.warn(
+                        EXCEPTION.getMarker(),
```
this marker should be `SOCKET_EXCEPTIONS`

kfa-aguda replied:
updated

lpetrovic05 commented:
discussed: add a test where the inbound and outbound handlers talk to each other. negative and positive case

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
```diff
@@ -380,8 +388,50 @@ public static boolean hasAnyCauseSuppliedType(
+     * @return info of the corresponding peer
+     * */
+    public static PeerInfo validateTLSPeer(
+            final @NonNull SSLSocket sslSocket, final @NonNull List<PeerInfo> peerInfoList) {
```
```suggestion
            @NonNull final SSLSocket sslSocket, @NonNull final List<PeerInfo> peerInfoList) {
```

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/Utilities.java:
I see these exact changes in this file on your other PR https://github.com/hashgraph/hedera-services/pull/12413

Not sure which PR builds off of the other, but if there is a dependency we should probably wait to do review on the second until the first has been fully reviewed and merged.

kfa-aguda replied:
put the other PR into draft. Will bring it back it after this is merged.

cody-littley reviewed (changes requested):
I'm unsure how to review this, there seem to be a lot of overlapping changes between this PR and https://github.com/hashgraph/hedera-services/pull/12413.

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkUtils.java:
```diff
@@ -171,4 +186,36 @@ public static String formatException(final Throwable e) {
+    /**
+     * identifies the client on the other end of the socket using their signing certificate.
+     *
+     * @param certs        the list of all certificates in the public keystore
```
this part of the javadoc is wrong, this is a the certificate chain presented my the peer through TLS

kfa-aguda replied:
updated. Moved it all into a separate class .

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkUtils.java:
```diff
@@ -171,4 +186,36 @@ public static String formatException(final Throwable e) {
+                        .getSubjectX500Principal()
+                        .equals(agreementCert.getIssuerX500Principal()))
+                .collect(Collectors.toSet());
+        final Optional<PeerInfo> peer = peers.stream().findFirst();
+        if (peer.isEmpty()) {
+            rateLimitedLogger.warn(
+                    SOCKET_EXCEPTIONS.getMarker(),
+                    "Unable to find a matching peer with the presented certificate {}.",
+                    agreementCert);
+        } else {
+            rateLimitedLogger.info(
+                    SOCKET_EXCEPTIONS.getMarker(),
+                    "Found {} matching peers for the presented certificate",
+                    peers.size());
+            matchingPair = peer.get();
```
this logic seems wrong. if we find a peer, it will log with a `SOCKET_EXCEPTIONS` marker. we should only log if there is more than one match. also, if this happens, we should use error level and `EXCEPTION` marker

kfa-aguda replied:
Looks like the logic is correct as per your comment, but the choice of logging isn't. Following up to clarify the markers.

edward-swirldslabs replied:
Just trying to follow the conversation.  There should be 3 rate limited logs here?  (1) error if there are no matches, (2) error if there are too many matches, (3) Info for successful single match?

edward-swirldslabs replied:
I don't know if my next comment is relevant here, but when using rate limited loggers, it is per message instance that needs to be rate limited independently of rate limiting other messages.     Suppose there are 3 log messages here for the different possible code-paths, and all 3 use the same rate limited logger.  Then the logging of a failure code-path will block the reporting of a successful code path until the time limit expires.  Vice versa, a report of success will block reporting of failures until the time limit expires.   All three code paths may need to have their own rate limited logger since we want to know the 1st instance of a message and only block out duplication of messages that have already been reported

kfa-aguda replied:
> Just trying to follow the conversation. There should be 3 rate limited logs here? (1) error if there are no matches, (2) error if there are too many matches, (3) Info for successful single match?

There should be two. Error if there are no matches, info if there are multiple matches.

Your second point is super helpful. I've just modified to add an additional rate limiter so one doesn't block the other. Unfortunately, I don't have a mechanism for testing/observing this behavior locally.

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -89,43 +80,24 @@ public InboundConnectionHandler(
-
-            dos.writeInt(ByteConstants.COMM_CONNECT); // send an ACK for creating connection
-            dos.flush();
+            final NodeId otherId = Objects.requireNonNull(connectedPeer).nodeId();
```
if the socket is not an SSLSocket this will throw an NPE, this is confusing, it should be handled differently. If we only support SSL sockets, then we should one use those

kfa-aguda replied:
I proposed getting rid of plain TCP as no one uses it today (as far as I can tell), and made this design assumption based on plain tcp connections going away (I have an unsubmitted PR for it).

If we want to support both, my preference will be to have two implementations of this class - one that figures out the other peer like before, and a new one that uses the new peer identification logic. The former will continue to require the address book to identify the connected peer, which re-introduces the separation of concern problem - one of the goals of isolating this network component. 

Another option of course is to throw a bunch of if else in this one class to support both. Not a fan but it may work. What do you think?

edward-swirldslabs replied:
Just a comment on code aesthetics.  Line 100 has 2 purposes, assert the invariant that the connectedPeer is not null and get the node id.   That first purpose feels like it should be closer to the definition of `connectedPeer`, like we wouldn't move forward into lines 95-98 if line 100 is going to fail as a result of the outcome of lines 95.    I think it feels easier to understand code when invariants are checked/enforced in proximity to the code which might violate them.

edward-swirldslabs replied:
`If we want to support both, my preference will be to have two implementations of this class`

An `InboundSSLConnectionHandler` and an `InboundNonSSLConnectionHandler`? 

If there is a lot of code duplication between the two classes, I'm in favor of it being a single class.   If the logic surrounding the if statements is large, duplicated between 2 (or more) classes, and needs to be updated, then we've created a scenario where code maintenance has to be made across all classes.    Keeping it one class with if statements minimizes the ancillary code duplication.    If there were special casing throughout the file for SSL vs NonSSL, then I'd say that might be a case for use of generic parameterized over the instance type.

lpetrovic05 replied:
The epic says we don't have to support plain TCP, getting rid of it is easier

kfa-aguda replied:
thanks @edward-swirldslabs. I was thinking more like an AbstractInboundConnectionHandler that keeps the common code and two concretes that handle SSL vs Non-SSL. Thankfully we don't have to go this route. Makes all the sense to just get rid of it as @lpetrovic05 says.

Long term, I'll revisit this code to actually make the inbound and outbound handlers share code. Right now, they do quite similar things but can benefit from improved polymorphism.

lpetrovic05 commented on platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/network/NetworkUtilsTest.java:
```diff
@@ -40,4 +59,42 @@ void handleNetworkExceptionTest() {
+            throws URISyntaxException, KeyLoadingException, KeyStoreException, InvalidAlgorithmParameterException {
+
+        // sample node names from mainnet
+        final List<String> names =
```
why just 10 nodes? why not all of them?

kfa-aguda replied:
There are 157 of those certificate entries in main-net. For this test, having a few peers get identified is sufficient to prove the logic works (since all the certificates are generated exactly the same way)

kfa-aguda replied:
I've updated the tests to use all of them

lpetrovic05 commented on platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/resources/com/swirlds/platform/crypto/agrCerts.data:
is this file used anywhere?

kfa-aguda replied:
deleted

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/connectivity/SocketFactoryTest.java:
```diff
@@ -67,38 +78,48 @@ class SocketFactoryTest {
         final Thread server = new Thread(() -> {
             try {
                 final Socket s = serverSocket.accept();
+                if (s instanceof SSLSocket) {
+                    if (peerInfoList.stream().anyMatch(peer -> mockingDetails(peer.signingCertificate())
+                            .isMock())) {
+                        // we've passed a mocked peer's certificate as an example of an invalid agreement
+                        // certificate - it possesses no valid issuer's principal
+                        assertNull(NetworkUtils.identifyTlsPeer(
+                                ((SSLSocket) s).getSession().getPeerCertificates(), peerInfoList));
+                    } else {
+                        assertNotNull(NetworkUtils.identifyTlsPeer(
+                                ((SSLSocket) s).getSession().getPeerCertificates(), peerInfoList));
+                    }
+                }
```
this looks like its beyond the scope of this test

kfa-aguda replied:
moved out to a dedicated test

kfa-aguda commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,14 +48,13 @@
  */
 public class InboundConnectionHandler {
     private static final Logger logger = LogManager.getLogger(InboundConnectionHandler.class);
+    private static final RateLimitedLogger rateLimitedLogger =
```
As per @edward-swirldslabs's previous comment, looks like we need one logger per spammy message, so i've introduced it.

kfa-aguda commented:
> discussed: add a test where the inbound and outbound handlers talk to each other. negative and positive case

as discussed offline, a positive and negative test cases of inbound handler is sufficient. Implemented and updated

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,83 +48,57 @@
      */
-    public void handle(final Socket clientSocket) {
-        SerializableDataInputStream dis = null;
-        SerializableDataOutputStream dos = null;
-        NodeId otherId = null;
+    public void handle(final Socket clientSocket, final List<PeerInfo> peerInfoList) {
```
Missing param documentation for peerInfoList

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,83 +48,57 @@
+            if (clientSocket instanceof final SSLSocket sslSocket) {
+                final PeerInfo connectedPeer = networkPeerIdentifier.identifyTlsPeer(
+                        sslSocket.getSession().getPeerCertificates(), peerInfoList);
+                otherId = Objects.requireNonNull(connectedPeer).nodeId();
```
identifyTlsPeer() returns null if it can't identify the peer, which means an unidentified peer will cause this this to throw a nullpointer exception. I'm wondering if it might be more appropriate to throw a more specific exception.

Also, we want to be careful to ensure that this failure will not generate excessive amounts of log spam (throwing the proper exception type plays into this). If we can't identify a peer, the amount of logging we do should be no more than if the connection breaks for any other reason.

kfa-aguda replied:
I think the logic here is that if we can't identify a peer, it's just regular business ... we log it and close the socket. The logging currently happens in NetworkPeerIdentifier#identifyTlsPeer, and the closing happens here when we throw the NPE. I agree it can be better though so I've changed it to:
`
            final SSLSocket sslSocket = ((SSLSocket) clientSocket);
            final PeerInfo connectedPeer = networkPeerIdentifier.identifyTlsPeer(
                    sslSocket.getSession().getPeerCertificates(), peerInfoList);

            if (connectedPeer == null) {
                sslSocket.close();
                return;
            }
            otherId = connectedPeer.nodeId();
`
This gets rid of the instanceOf check and fails if we can't cast (we won't because it's always an ssl socket), and also handles the null check more clearly.
(We are getting rid of Plain tcp soon, so the argument to this method will change to SSLSocket instead of Socket)

cody-littley commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/NetworkPeerIdentifierTest.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    @Test
+    @DisplayName(
+            "asserts that when none of the peer's certificate matches any of the certs in the trust store, identifyTlsPeer returns null")
```
This seems more appropriate as a javadoc instead of as a display name

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkPeerIdentifier.java:
```diff
@@ -0,0 +1,83 @@
+    public @Nullable PeerInfo identifyTlsPeer(
+            final @NonNull Certificate[] certs, final @NonNull List<PeerInfo> peerInfoList) {
+        PeerInfo matchingPair = null;
+        final X509Certificate agreementCert = (X509Certificate) certs[0];
```
we should check the length of the array first

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,83 +48,57 @@
             clientSocket.setTcpNoDelay(socketConfig.tcpNoDelay());
             clientSocket.setSoTimeout(socketConfig.timeoutSyncClientSocket());
 
+            if (clientSocket instanceof final SSLSocket sslSocket) {
```
we should close the socket if we cant identify the peer

cody-littley replied:
If we can, we should refactor this so that we don't rely on an `instanceof` check to make this type of decision. Perhaps when we construct this object, we can configure things differently depending on if we expect SSL sockets or not.

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkPeerIdentifier.java:
```diff
@@ -0,0 +1,83 @@
+            final @NonNull Certificate[] certs, final @NonNull List<PeerInfo> peerInfoList) {
+        PeerInfo matchingPair = null;
+        final X509Certificate agreementCert = (X509Certificate) certs[0];
+        final Set<PeerInfo> peers = peerInfoList.stream()
+                .filter(peerInfo -> ((X509Certificate) peerInfo.signingCertificate())
+                        .getSubjectX500Principal()
+                        .equals(agreementCert.getIssuerX500Principal()))
+                .collect(Collectors.toSet());
```
Instead of iterating over all peers each time, could we instead store this in formation in a map where the `agreementCert.getIssuerX500Principal()` is the key?

kfa-aguda replied:
discussed offline - this utility is intended to be stateless plus the performance benefit of making it stateful is debatable. We will discuss further with @lpetrovic05 Monday.

lpetrovic05 replied:
I don't think there is much of a performance benefit, but I also think its ok to go to a map lookup. Either way is good with me

kfa-aguda replied:
i really think this utility should stay fully stateless.

kfa-aguda replied:
Raised https://github.com/hashgraph/hedera-services/issues/12709 as immediate follow up.

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkPeerIdentifier.java:
```diff
@@ -0,0 +1,83 @@
+    public @Nullable PeerInfo identifyTlsPeer(
+            final @NonNull Certificate[] certs, final @NonNull List<PeerInfo> peerInfoList) {
+        PeerInfo matchingPair = null;
+        final X509Certificate agreementCert = (X509Certificate) certs[0];
```
I suggest adding documentation describing why `certs[0]` is special. This was initially confusing until I dug into documentation.

kfa-aguda replied:
Really? Seems clear enough as it's right there in `SSLSession#getPeerCertificates` API doc. I've copied it here too for clarity.

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,83 +48,57 @@
-        SerializableDataOutputStream dos = null;
-        NodeId otherId = null;
+    public void handle(final Socket clientSocket, final List<PeerInfo> peerInfoList) {
         long acceptTime = 0;
+        NodeId otherId = null;
```
If these are moved into the `try` block then they can be declared as final variables.

kfa-aguda replied:
the orderId, yes, but not the acceptTime as it's needed in the catch blocks.

kfa-aguda replied:
resolving as discussed

kfa-aguda commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkPeerIdentifier.java:
```diff
@@ -0,0 +1,83 @@
+    /**
+     * limits the frequency of error log statements
+     */
+    private static final RateLimitedLogger noPeerFoundLogger =
+            new RateLimitedLogger(logger, Time.getCurrent(), Duration.ofMinutes(5));
+
+    private static final RateLimitedLogger multiplePeersFoundLogger =
+            new RateLimitedLogger(logger, Time.getCurrent(), Duration.ofMinutes(5));
```
I'll change this. It's super confusing though as there are a tonne of places in production code using Time.getCurrent() at the moment. so it does feel like a adhoc standard. We should probably log a ticket to address this sometime.

kfa-aguda replied:
Discussed with Cody offline. The Hashgraph team is sticking to this new pattern.

kfa-aguda commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/ConnectionServer.java:
```diff
@@ -44,36 +46,36 @@ public class ConnectionServer implements InterruptableRunnable {
             final ThreadManager threadManager,
             final int port,
             final SocketFactory socketFactory,
-            final Consumer<Socket> newConnectionHandler) {
+            final BiConsumer<Socket, List<PeerInfo>> newConnectionHandler,
+            final List<PeerInfo> peerInfoList) {
```
This class is in the middle of a big rework. This PR only adds one more parameter, so it does make sense to defer back-porting the annotations to a change that's much more involving. That other change is in draft, should follow this PR.

poulok commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,83 +48,57 @@
         long acceptTime = 0;
+        NodeId otherId = null;
         try {
             acceptTime = System.currentTimeMillis();
```
Please use `Time.currentTimeMillis()`

poulok commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/NetworkPeerIdentifierTest.java:
```diff
@@ -0,0 +1,134 @@
+        final Certificate[] certificates =
+                trustAnchors.stream().map(TrustAnchor::getTrustedCert).toArray(Certificate[]::new);
+        final PeerInfo matchedPeer = new NetworkPeerIdentifier().identifyTlsPeer(certificates, peerInfoList);
+        Assertions.assertNotNull(matchedPeer);
```
If the method under test simply returned a random or incorrect `PeerInfo` from the list provided, this test would pass. Best to add more assertions to verify that the correct peer was returned.

kfa-aguda replied:
I agree it's generally correct to assert specific properties of an object in tests but we can't do that here as:
- the certificates loaded from the key file are unordered
- we return the first match, not the only match.

kfa-aguda replied:
Got it. Changed to actually match each node, put them into a set, and assert the final set has the exact same elements as the original list of peers.

kfa-aguda replied:
and also added one more test to identify a specific node.

poulok commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/NetworkPeerIdentifierTest.java:
```diff
@@ -0,0 +1,134 @@
+                List.of("node1", "node2", "node3", "node4", "node5", "node6", "node7", "node8", "node9", "node10");
+        // sample pfx file grabbed from mainnet
+        final KeyStore publicKeys = loadKeys(
+                ResourceLoader.getFile("preGeneratedKeysAndCerts/").resolve("publicMainnet.pfx"),
```
Does this file contain public keys for a single mainnet node or multiple? Either way, it is helpful to include that information here to aid readers in understanding exactly what this test is doing.

kfa-aguda replied:
multiple.
The javadoc on the test says something to that effect
`    /**
     * Tests that given a list of valid Swirlds production certificates (like the type used in mainnet),
     * {@link NetworkPeerIdentifier#identifyTlsPeer(Certificate[], List)} is able to successfully identify a matching
     * peer.
     */`
I can add `we've grabbed the public keys for all nodes in mainnet for this test` - will that help? If not, I'd take suggestions :)

poulok replied:
shame on me for not reading the javadoc! It's fine as is

poulok commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/NetworkPeerIdentifierTest.java:
```diff
@@ -0,0 +1,134 @@
+    void testIdentifyTlsPeerReturnsNull()
+            throws URISyntaxException, KeyLoadingException, KeyStoreException, NoSuchAlgorithmException,
+                    NoSuchProviderException, KeyGeneratingException {
+        final List<String> names =
+                List.of("node1", "node2", "node3", "node4", "node5", "node6", "node7", "node8", "node9", "node10");
+        // sample pfx file grabbed from mainnet
+        final KeyStore publicKeys = loadKeys(
+                ResourceLoader.getFile("preGeneratedKeysAndCerts/").resolve("publicMainnet.pfx"),
+                "password".toCharArray());
+        final PublicStores publicStores = PublicStores.fromAllPublic(publicKeys, names);
+
+        // create list of peers with valid certs
+        final List<PeerInfo> peerInfoList = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            final String name = names.get(i);
+            final NodeId node = new NodeId(i);
+            final PeerInfo peer = new PeerInfo(
+                    node,
+                    names.get(i),
+                    "127.0.0.1",
+                    Objects.requireNonNull(publicStores.getCertificate(KeyCertPurpose.SIGNING, name)));
+            peerInfoList.add(peer);
+        }
```
This is duplicated from the previous test. I think it would be cleaner to have a setup method that performs these tasks and makes the required variables available to all tests in the class. Then it is very easy for reader to see what is unique to each test, with the added bonus of reduced line count.

kfa-aguda replied:
agreed. Fixed

kfa-aguda commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/connectivity/InboundConnectionHandler.java:
```diff
@@ -50,83 +48,57 @@
     public InboundConnectionHandler(
             @NonNull final PlatformContext platformContext,
             @NonNull final ConnectionTracker connectionTracker,
+            @NonNull final NetworkPeerIdentifier networkPeerIdentifier,
             @NonNull final NodeId selfId,
-            @NonNull final AddressBook addressBook,
             @NonNull final InterruptableConsumer<Connection> newConnectionConsumer,
-            final boolean doVersionCheck,
-            @NonNull final SoftwareVersion softwareVersion,
             @NonNull final Time time) {
```
done

lpetrovic05 commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/NetworkPeerIdentifierTest.java:
```diff
@@ -0,0 +1,139 @@
+                CryptoStatic.generateCertificate(name, rsaKeyPair1, name, rsaKeyPair1, secureRandom);
+        final X509Certificate ecCert =
+                CryptoStatic.generateCertificate(name, rsaKeyPair2, name, rsaKeyPair2, secureRandom);
+        final Certificate[] certificates = new Certificate[] {rsaCert, ecCert};
```
why are we passing in 2 certificates here?

cody-littley commented on platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/network/NetworkPeerIdentifier.java:
```diff
@@ -0,0 +1,97 @@
+
+    private final RateLimitedLogger multiplePeersFoundLogger;
+
+    public NetworkPeerIdentifier(@NonNull final PlatformContext platformContext) {
```
please add a javadoc

kfa-aguda replied:
Will take care in another PR

cody-littley commented on platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/network/connectivity/InboundConnectionHandlerTest.java:
```diff
@@ -0,0 +1,137 @@
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.Mockito;
+
+class InboundConnectionHandlerTest extends ConnectivityTestBase {
```
When I run these tests locally, I notice that the number of certs used is exactly 1. This test should have a much larger set of peers... identifying a peer when there is only one peer doesn't really exercise this code the way it needs to be exercised.

cody-littley replied:
For the other tests that have a list of peers we should also ensure we have more than one peer (I did not check).

kfa-aguda replied:
we discussed this offline. Agreed to close.

cody-littley reviewed (approved):
We discussed, an agreement was made to address open issue as an immediate follow up PR which will be merged before the 0.50 branch is cut.
